stan/math/rev/core/vari.hpp

```diff
@@ -179,11 +177,393 @@ class vari_value<T, std::enable_if_t<std::is_floating_point<T>::value>>
 using vari = vari_value<double>;
 
+/**
+ * The variable implementation for Eigen dense matrix types.
+ *
+ * This class is complete (not abstract) and may be used for
+ * constants.
+ *
+ * A variable implementation is constructed with a constant
+ * value. It also stores the adjoint for storing the partial
+ * derivative with respect to the root of the derivative tree.
+ *
+ */
+template <typename T>
+class vari_value<T, std::enable_if_t<is_eigen_dense_base<T>::value>>
+    : public vari_base {
+ public:
+  static_assert(
+      is_plain_type<T>::value,
+      "The template for this var is an"
+      " expression but a var_value's inner type must be assignable such as"
+      " a double, Eigen::Matrix, or Eigen::Array");
+
+  /**
+   * `PlainObject` represents a user constructible type such as Matrix or Array
+   */
+  using PlainObject = std::decay_t<plain_type_t<T>>;
+  using Scalar = PlainObject;  // The underlying type for this class
+  using value_type = Scalar;   // The underlying type for this class
+  using eigen_scalar = value_type_t<PlainObject>;  // A floating point type
+  using eigen_map = Eigen::Map<PlainObject>;       // Maps for adj_ and val_
+  using vari_type
+      = vari_value<T, std::enable_if_t<is_eigen_dense_base<T>::value>>;
+  eigen_scalar* val_mem_;  // Pointer to memory allocated on the stack for val_
+  eigen_scalar* adj_mem_;  // Pointer to memory allocated on the stack for adj_
+  /**
+   * Number of rows known at compile time
+   */
+  static constexpr Eigen::Index RowsAtCompileTime = Scalar::RowsAtCompileTime;
+  /**
+   * Number of columns known at compile time
+   */
+  static constexpr Eigen::Index ColsAtCompileTime = Scalar::ColsAtCompileTime;
+
+  /**
+   * The adjoint of this variable, which is the partial derivative
+   * of this variable with respect to the root variable.
+   */
+  eigen_map adj_;
+
+  /**
+   * The value of this variable.
+   */
+  const eigen_map val_;
+
+  /**
+   * Construct a dense Eigen variable implementation from a value. The
+   * adjoint is initialized to zero.
+   *
+   * All constructed variables are added to the stack. Variables
+   * should be constructed before variables on which they depend
+   * to insure proper partial derivative propagation.  During
+   * derivative propagation, the chain() method of each variable
+   * will be called in the reverse order of construction.
+   *
+   * @tparam S A dense Eigen type that is convertible to `value_type`
+   * @param x Value of the constructed variable.
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr>
+  explicit vari_value(S&& x)
+      : val_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        adj_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        adj_(make_adj(x)),
+        val_(make_val(x)) {
+    ChainableStack::instance_->var_stack_.push_back(this);
```

Comment:
How about we completely eliminate these private functions?

```suggestion
        adj_(adj_mem_, x.rows(), x.cols()),
        val_(val_mem_, x.rows(), x.cols()) {
    adj_.setZero();
    val_ = x;
    ChainableStack::instance_->var_stack_.push_back(this);
```

Reply:
Yes I figured out how to do this over the weekend!